invenio_indexer/api.py

```diff
@@ -40,13 +42,28 @@
 
 
 def _record_to_index(record):
-    """Get index/doctype given a record."""
-    index, doctype = schema_to_index(record.get('$schema', ''))
-    if index and doctype:
-        return index, doctype
+    """Get index/doc_type given a record."""
+    index_names = current_search.mappings.keys()
+    schema = record.get('$schema', '')
+    if isinstance(schema, dict):
+        schema = schema.get('$ref', '')
+
+    index, doc_type = schema_to_index(schema, index_names=index_names)
+
+    if index and doc_type:
+        return index, doc_type
     else:
-        return current_app.config['INDEXER_DEFAULT_INDEX'], \
-            current_app.config['INDEXER_DEFAULT_DOCTYPE']
+        return (current_app.config['INDEXER_DEFAULT_INDEX'],
+                current_app.config['INDEXER_DEFAULT_DOC_TYPE'])
+
+
+class Producer(KombuProducer):
+    """Producer validating published messages."""
+
+    def publish(self, data, **kwargs):
```

Comment:
Would it perhaps be better force the signature to ``publish(id, op, index, doctype)`` to make the API more rigid? Just thinking that if we expose the ``create_producer()`` as part of the API, then there's quite a chance to pass something wrong into the queue.